core/src/main/java/org/apache/spark/memory/TaskMemoryManager.java

```diff
@@ -256,7 +261,20 @@ public MemoryBlock allocatePage(long size, MemoryConsumer consumer) {
       }
       allocatedPages.set(pageNumber);
     }
-    final MemoryBlock page = memoryManager.tungstenMemoryAllocator().allocate(acquired);
+    MemoryBlock page = null;
+    try {
+      page = memoryManager.tungstenMemoryAllocator().allocate(acquired);
+    } catch (OutOfMemoryError e) {
+      logger.warn("Failed to allocate a page ({} bytes), try again.", acquired);
+      // there is no enough memory actually, it means the actual free memory is smaller than
+      // MemoryManager thought, we should keep the acquired memory.
+      acquiredButNotUsed += acquired;
+      synchronized (this) {
+        allocatedPages.clear(pageNumber);
+      }
+      // this could trigger spilling to free some pages.
+      return allocatePage(size, consumer);
```

Comment:
Since we are continue hold some memory, the amount of free memory should become smaller and smaller, it will fail to acquire soon.

yes, it's better to move `acquiredButNotUsed += acquired` into the synchronized sections. If acquiredButNotUsed is not calculated correctly (because risk conditions), you will only saw an warning message at the end of a task.